dns_sd_avahi.c

```diff
@@ -14,42 +15,191 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
+ * Some of this is insipred from libavahi's example:
+ * https://avahi.org/doxygen/html/client-browse-services_8c-example.html
+ * which is also LGPL 2.1 or later.
+ *
  * */
 
-#include "iio.h"
+#include "iio-private.h"
 
 #include <errno.h>
 #include <stdbool.h>
 #include <string.h>
 #include <time.h>
+#include <sys/time.h>
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
 
 #include <avahi-common/error.h>
 #include <avahi-common/simple-watch.h>
+#include <avahi-common/malloc.h>
 #include <avahi-client/client.h>
 #include <avahi-client/lookup.h>
 
 #include "debug.h"
 
+/* in network.c */
+int create_socket(const struct addrinfo *addrinfo, unsigned int timeout);
+#define DEFAULT_TIMEOUT_MS 5000
+#define IIOD_PORT 30431
+
+
 struct avahi_discovery_data {
 	AvahiSimplePoll *poll;
 	AvahiAddress *address;
-	uint16_t *port;
-	bool found, resolved;
+	char addr_str[AVAHI_ADDRESS_STR_MAX];
+	char *hostname;
+	uint16_t found, resolved;
+	uint16_t port;
+	struct avahi_discovery_data *next;
 };
 
+/*
+ * Fundamentally, this builds up a linked list to manage
+ * potential clients on the network
+ */
+
+static struct avahi_discovery_data * new_discovery_data(void)
+{
+	struct avahi_discovery_data *data;
+
+	data = zalloc(sizeof(struct avahi_discovery_data));
+	if (!data)
+		return NULL;
+	data->address = zalloc(sizeof(struct AvahiAddress));
+	if (!data->address)
```

Comment:
this return path should probably `free(data)`;

Reply:
Thanks - good catch